cdap-app-templates/cdap-etl/cdap-data-pipeline-base/src/main/java/io/cdap/cdap/datapipeline/service/RemoteConnectionTaskBase.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package io.cdap.cdap.datapipeline.service;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+import io.cdap.cdap.api.artifact.ArtifactId;
+import io.cdap.cdap.api.data.schema.Schema;
+import io.cdap.cdap.api.macro.MacroEvaluator;
+import io.cdap.cdap.api.macro.MacroParserOptions;
+import io.cdap.cdap.api.service.http.ServicePluginConfigurer;
+import io.cdap.cdap.api.service.worker.RunnableTask;
+import io.cdap.cdap.api.service.worker.RunnableTaskContext;
+import io.cdap.cdap.api.service.worker.SystemAppTaskContext;
+import io.cdap.cdap.datapipeline.connection.DefaultConnectorConfigurer;
+import io.cdap.cdap.datapipeline.connection.DefaultConnectorContext;
+import io.cdap.cdap.etl.api.connector.Connector;
+import io.cdap.cdap.etl.api.connector.ConnectorConfigurer;
+import io.cdap.cdap.etl.api.connector.ConnectorContext;
+import io.cdap.cdap.etl.api.connector.ConnectorSpec;
+import io.cdap.cdap.etl.api.connector.ConnectorSpecRequest;
+import io.cdap.cdap.etl.common.ArtifactSelectorProvider;
+import io.cdap.cdap.etl.common.BasicArguments;
+import io.cdap.cdap.etl.common.DefaultMacroEvaluator;
+import io.cdap.cdap.etl.common.OAuthMacroEvaluator;
+import io.cdap.cdap.etl.common.SecureStoreMacroEvaluator;
+import io.cdap.cdap.etl.proto.ArtifactSelectorConfig;
+import io.cdap.cdap.etl.proto.connection.Connection;
+import io.cdap.cdap.etl.proto.connection.ConnectorDetail;
+import io.cdap.cdap.etl.proto.connection.PluginDetail;
+import io.cdap.cdap.etl.proto.connection.PluginInfo;
+import io.cdap.cdap.etl.proto.connection.SpecGenerationRequest;
+import io.cdap.cdap.etl.proto.validation.SimpleFailureCollector;
+import io.cdap.cdap.etl.spec.TrackedPluginSelector;
+import io.cdap.cdap.internal.io.SchemaTypeAdapter;
+
+import java.nio.charset.StandardCharsets;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Base class for {@link RunnableTask} for connection handling
+ */
+public abstract class RemoteConnectionTaskBase implements RunnableTask {
+
+  private static final Gson GSON = new Gson();
+
+  @Override
+  public void run(RunnableTaskContext context) throws Exception {
+    //Get SystemAppTaskContext
+    SystemAppTaskContext systemAppContext = context.getRunnableTaskSystemAppContext();
+    execute(context, systemAppContext);
+  }
+
+  /**
+   * execute method for child class to implement
+   *
+   * @param context              {@link RunnableTaskContext}
+   * @param systemAppTaskContext {@link SystemAppTaskContext}
+   * @throws Exception
+   */
+  protected abstract void execute(RunnableTaskContext context,
```

Comment:
I feel we can put most of the logic on creating configurer, context here. And each individual class only need to run the code inside the `try (Connector ...)` block, this will remove most duplication in the task class

Reply:
Moved out common code to base class